### PR TITLE
Rob Youngman's changes 

### PR DESCRIPTION
**Pull request by youngmanr**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,3 +1,74 @@
...
+2.2.3 :002 > require './lib/messenger'
+ => true
+2.2.3 :003 > require './lib/takeaway'
+ => true
+2.2.3 :004 > require './lib/restaurant'
+ => true
+2.2.3 :005 > messenger = Messenger.new
+ => #<Messenger:0x007f8fe2be7ea8>
+2.2.3 :006 > menu = Menu.new
+ => #<Menu:0x007f8fe2bdcb98 @items={"Pepperoni"=>1, "Hot Shot"=>1, "Jardin"=>2, "Provencale"=>2, "Roquefort"=>3, "Neptune"=>3, "Poulet"=>4, "Four Seasons"=>4}>
```
Comment (DeathRay1977) [inline README.md]: These instructions could have been a little clearer. Some reformatting maybe?

File: spec/takeaway_spec.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
+require 'takeaway'
+
+describe Takeaway do
+  let(:menu) {double :menu}
+  let(:restaurant) {double :restaurant}
+  subject(:takeaway) {described_class.new( :restaurant, :menu)}
```
Comment (DeathRay1977) [inline spec/takeaway_spec.rb]: No colons needed in new()

Comment (d9nny) [inline spec/takeaway_spec.rb]: Need to ref the double instead of using a symbol - for restaurant & menu

File: spec/takeaway_spec.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+      expect{ takeaway.add('Neptune') }.to change { takeaway.basket.length }.by 1
+    end
+
+    it 'add second item' do
+      allow(menu).to receive(:contains?).with('Four Seasons').and_return(true)
+      expect { takeaway.add('Four Seasons') }.to change { takeaway.basket.length }.by 1
+    end
+
+    it 'fails to add unknown item to basket' do
+    allow(menu).to receive(:contains?).with('Tandoori').and_return(true)
```
Comment (d9nny) [inline spec/takeaway_spec.rb]: change return to false

File: lib/takeaway.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
+class Takeaway
+  attr_reader :basket
+
+  def initialize( restaurant, menu)
+    @menu = menu
+    @restaurant = restaurant
+    @basket = {}
```
Comment (d9nny) [inline lib/takeaway.rb]: Could use Hash.new(0)

File: lib/messenger.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
+require 'twilio-ruby'
+
+class Messenger
+  def create( from, to, body)
+    account_sid = 'AC89d950d129482b828259244e80366b30'
+    auth_token = ''
+    # set up a client to talk to the Twilio REST API
+    @client = Twilio::REST::Client.new account_sid, auth_token
```
Comment (danldb) [inline lib/messenger.rb]: you could inject the client into the method to allow easy stubbing - also means if you ever change your SMS provider it is easy to replace one for the other.

File: lib/restaurant.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
+
+class Restaurant
+  def initialize( messenger, menu)
+    @messenger = messenger
+    @menu = menu
+    @header = { from: '+441290211165', to: '+447527994300'}
+  end
+
+  def request_takeaway
+    takeaway = Takeaway.new(self, @menu)
```
Comment (danldb) [inline lib/restaurant.rb]: this completely couples Takeaway and Restaurant - you cannot have one without the other. This makes it harder to reuse or maintain your code. Use DI instead.

File: lib/takeaway.rb [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    puts @menu
+  end
+
+  def summary
+    str = ""
+    basket.each do |name, qty|
+      price = @menu.price( name)
+      str = str + "#{name} " + "x#{qty} = £#{price * qty} "
+    end
+    puts "#{str}"
```
Comment (danldb) [inline lib/takeaway.rb]: don't use puts, don't return a string. We did not ask for an interface.

File: spec/menu_spec.rb [hunk 8]
```diff
@@ -0,0 +1,9 @@
+require 'menu'
+
+describe Menu do
+  subject(:menu) {described_class.new}
+
+  it { is_expected.to respond_to(:contains?).with(1).argument }
+  it { is_expected.to respond_to(:price).with(1).argument }
+
```
Comment (danldb) [inline spec/menu_spec.rb]: not testing the actions of the methods.

File: spec/messenger_spec.rb [hunk 9]
```diff
@@ -0,0 +1,8 @@
+require 'messenger'
+
+describe Messenger do
+  subject(:Restaurant) {described_class.new}
+
+  it { is_expected.to respond_to(:create) }
```
Comment (danldb) [inline spec/messenger_spec.rb]: This is not testing the action, only the method name.

File: spec/takeaway_spec.rb [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  let(:menu) {double :menu}
+  let(:restaurant) {double :restaurant}
+  subject(:takeaway) {described_class.new( restaurant, menu)}
+
+  it { is_expected.to respond_to(:add).with(1).argument }
+  it { is_expected.to respond_to(:remove).with(1).argument }
+  it { is_expected.to respond_to(:menu) }
+  it { is_expected.to respond_to(:summary) }
+  it { is_expected.to respond_to(:total) }
+
```
Comment (danldb) [inline spec/takeaway_spec.rb]: Lots of redundant respond_to tests.

**Comment by danldb** [comment]
:checkered_flag: 

Well done, good submission. See inline notes for refactoring advice.

**Pull request closed without merging**